satellite/gracefulexit/endpoint.go

```diff
@@ -221,15 +223,49 @@ func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
 
 	for {
 		pendingCount := pending.length()
+
 		// if there are no more transfers and the pending queue is empty, send complete
 		if atomic.LoadInt32(&morePiecesFlag) == 0 && pendingCount == 0 {
-			// TODO check whether failure threshold is met before sending completed
-			// TODO needs exit signature
-			transferMsg := &pb.SatelliteMessage{
-				Message: &pb.SatelliteMessage_ExitCompleted{
-					ExitCompleted: &pb.ExitCompleted{},
-				},
+			exitStatusRequest := &overlay.ExitStatusRequest{
+				NodeID:         nodeID,
+				ExitFinishedAt: time.Now().UTC(),
+			}
+
+			progress, err := endpoint.db.GetProgress(ctx, nodeID)
+			if err != nil && !errs.Is(err, sql.ErrNoRows) {
+				return Error.Wrap(err)
 			}
+
+			var transferMsg *pb.SatelliteMessage
+			processed := progress.PiecesFailed + progress.PiecesTransferred
+			// check node's exiting progress to see if it has failed passed max failure threshold
+			if processed > 0 && progress.PiecesFailed > 0 &&
```

Comment:
nit - calculate percentage failed outside of the if statement to make this easier to read.